include/rect.h

```diff
@@ -29,33 +31,79 @@ namespace DosBox {
 struct Rect {
 	constexpr Rect() = default;
 
-	constexpr Rect(const float _w, const float _h)
-	        : x(0.0f),
-	          y(0.0f),
-	          w(_w),
-	          h(_h)
-	{}
-
-	constexpr Rect(const int _w, const int _h)
-	        : x(static_cast<float>(0.0f)),
-	          y(static_cast<float>(0.0f)),
-	          w(static_cast<float>(_w)),
-	          h(static_cast<float>(_h))
-	{}
 	constexpr Rect(const float _x, const float _y, const float _w, const float _h)
 	        : x(_x),
 	          y(_y),
 	          w(_w),
 	          h(_h)
+	{
+		assert(w >= 0.0f);
+		assert(h >= 0.0f);
+	}
+
+	constexpr Rect(const float _w, const float _h)
+	        : Rect(0.0f, 0.0f, _w, _h)
 	{}
 
 	constexpr Rect(const int _x, const int _y, const int _w, const int _h)
-	        : x(static_cast<float>(_x)),
-	          y(static_cast<float>(_y)),
-	          w(static_cast<float>(_w)),
-	          h(static_cast<float>(_h))
+	        : Rect(static_cast<float>(_x), static_cast<float>(_y),
+	               static_cast<float>(_w), static_cast<float>(_h))
 	{}
 
+	constexpr Rect(const int _w, const int _h) : Rect(0, 0, _w, _h) {}
+
+	constexpr float x1() const
+	{
+		return x;
+	}
+
+	constexpr float y1() const
+	{
+		return y;
+	}
+
+	constexpr float x2() const
+	{
+		return x + w;
+	}
+
+	constexpr float y2() const
+	{
+		return y + h;
+	}
+
+	constexpr float cx() const
+	{
+		return x + w / 2;
```

Comment:
Suggest asserting that `w != 0`.

Reply:
Actually, not needed, we're not dividing by it.